vector/src/main/java/im/vector/app/features/onboarding/RegistrationActionHandler.kt

```diff
@@ -16,26 +16,80 @@
 
 class RegistrationActionHandler @Inject constructor() {
 
     suspend fun handleRegisterAction(registrationWizard: RegistrationWizard, action: RegisterAction): RegistrationResult {
         return when (action) {
-            RegisterAction.StartRegistration               -> registrationWizard.getRegistrationFlow()
-            is RegisterAction.CaptchaDone                  -> registrationWizard.performReCaptcha(action.captchaResponse)
-            is RegisterAction.AcceptTerms                  -> registrationWizard.acceptTerms()
-            is RegisterAction.RegisterDummy                -> registrationWizard.dummy()
-            is RegisterAction.AddThreePid                  -> registrationWizard.addThreePid(action.threePid)
-            is RegisterAction.SendAgainThreePid            -> registrationWizard.sendAgainThreePid()
-            is RegisterAction.ValidateThreePid             -> registrationWizard.handleValidateThreePid(action.code)
-            is RegisterAction.CheckIfEmailHasBeenValidated -> registrationWizard.checkIfEmailHasBeenValidated(action.delayMillis)
-            is RegisterAction.CreateAccount                -> registrationWizard.createAccount(action.username, action.password, action.initialDeviceName)
+            RegisterAction.StartRegistration               -> resultOf { registrationWizard.getRegistrationFlow() }
+            is RegisterAction.CaptchaDone                  -> resultOf { registrationWizard.performReCaptcha(action.captchaResponse) }
+            is RegisterAction.AcceptTerms                  -> resultOf { registrationWizard.acceptTerms() }
+            is RegisterAction.RegisterDummy                -> resultOf { registrationWizard.dummy() }
+            is RegisterAction.AddThreePid                  -> handleAddThreePid(registrationWizard, action)
+            is RegisterAction.SendAgainThreePid            -> resultOf { registrationWizard.sendAgainThreePid() }
+            is RegisterAction.ValidateThreePid             -> resultOf { registrationWizard.handleValidateThreePid(action.code) }
+            is RegisterAction.CheckIfEmailHasBeenValidated -> handleCheckIfEmailIsValidated(registrationWizard, action.delayMillis)
+            is RegisterAction.CreateAccount                -> resultOf {
+                registrationWizard.createAccount(
+                        action.username,
+                        action.password,
+                        action.initialDeviceName
+                )
+            }
         }
     }
+
+    private suspend fun handleAddThreePid(wizard: RegistrationWizard, action: RegisterAction.AddThreePid): RegistrationResult {
+        return runCatching { wizard.addThreePid(action.threePid) }.fold(
+                onSuccess = { it.toRegistrationResult() },
+                onFailure = {
+                    when {
+                        action.threePid is RegisterThreePid.Email && it.is401() -> RegistrationResult.SendEmailSuccess(action.threePid.email)
```

Comment:
the `401 == successful pid sent` logic is no longer handled at the UI/fragment layer but instead at our business logic layer 

This logic feels like something that should be part of the SDK